### ISSUE TITLE
Inserting and registering objects in two separate steps leads to a somewhat odd API

### ISSUE THREAD

**Issue by YorickPeterse** [stjepang/polling#16]
The current API of polling is (to my understanding) setup such that it requires two steps to poll an object:

1. Add the file descriptor/socket/etc using `Poller::insert()`
1. Register an interest using `Poller::interest()`

`Poller::insert()` also seems to be fiddling around with the non-blocking mode of the descriptor it's given, even if the descriptor is already in non-blocking mode.

For [Inko](https://inko-lang.org/) I'm using a custom-setup (see the files in [this directory](https://gitlab.com/inko-lang/inko/-/tree/master/vm/src/network_poller), built on the same underlying technology: epoll, kqueue, wepoll, etc. I'd like to use the polling crate instead for the following reasons:

1. I don't have to implement code that's already available in a crate
1. I want to be able to terminate polling threads cleanly. My implementation doesn't allow for this, while polling does (from what I understand at least)
1. I can, if necessary, help with this particular crate; instead of building my own thing that can't be reused outside of Inko

The way I use the poller is not compatible with what Poller does: when a non-blocking operation is performed, the file descriptor is registered (with its interest) in a single call. For example, for epoll() this translates to a single epoll_ctl() that registers an event with all the necessary flags (e.g. EPOLLIN) already set. I went with this approach because it requires fewer system calls, and makes the API easier to use. This brings me to my first question:

What is the motivation for having both `Poller::insert()` and `Poller::interest()`, instead of merging them into a single method? And would there be interest in having a method that combines these two methods, but using a single system call?

This ties in to my second question: why is Poller messing with the non-blocking status of a file descriptor? This introduces redundant system call overhead when the input descriptors are already in non-blocking mode, with no way of avoiding this. Would it be an option to perhaps add a `Poller::insert_nonblocking` method that just takes the descriptor as-is? That way one can opt-out of the additional system calls.

p.s. Sorry for the somewhat vague title, I couldn't come up with anything better.

**Comment by stjepang** [comment]
> What is the motivation for having both `Poller::insert()` and `Poller::interest()`, instead of merging them into a single method? And would there be interest in having a method that combines these two methods, but using a single system call?

It's because of the difference between EPOLL_CTL_ADD and EPOLL_CTL_MOD. It's annoying - I wish epoll didn't have this distinction but it does.

If we merged `insert()` and `interest()` into a single method, would that method take some kind of `bool` parameter indicating whether it's the first insert versus a modification?

> This ties in to my second question: why is Poller messing with the non-blocking status of a file descriptor? This introduces redundant system call overhead when the input descriptors are already in non-blocking mode, with no way of avoiding this.

It's just to make things a bit more ergonomic. It should be possible to remove one syscall easily - there's no point in setting O_NONBLOCK if it's already set. Still, I agree it would be good to optimize this.

I think instead of adding new oddly named methods, we should just do whatever breaking changes are necessary and then publish v2.0. Still, I would prefer to keep it as simple as possible...

**Comment by YorickPeterse** [comment]
@stjepang 

> If we merged insert() and interest() into a single method, would that method take some kind of bool parameter indicating whether it's the first insert versus a modification?

This is basically what I did: I have a `register_or_reregister` method that takes a `update` boolean. This is then wrapped by two separate `register` and `reregister` methods. I then track if a socket is registered in the socket itself.

Funny enough, as I read the above line I realise the Poller setup may not be an issue after all. I already track per socket if it's registered or not, meaning I can decide if a `Poller::insert()` is necessary or not. :smile: 

> I think instead of adding new oddly named methods, we should just do whatever breaking changes are necessary and then publish v2.0. Still, I would prefer to keep it as simple as possible

One option is to check the existing flags and not set O_NONBLOCK if already set. But this still requires a syscall.

The alternative is to expect sockets and such are already set in non-blocking mode, and bump to 2.0. However, I can see this being a bit annoying for users depending on how they use `Poller`. I do think that from a technical point of view this is probably the simplest change. If you're OK with taking this path, I'll gladly set up a pull request to make the necessary changes.

**Comment by stjepang** [comment]
> Funny enough, as I read the above line I realise the Poller setup may not be an issue after all. I already track per socket if it's registered or not, meaning I can decide if a `Poller::insert()` is necessary or not. 😄

So how about the following API?

```
impl Poller {
    pub fn insert(&self, source: impl Source, interest: Event) -> io::Result<()>;
    pub fn update(&self, source: impl Source, interest: Event) -> io::Result<()>;
    pub fn remove(&self, source: impl Source) -> io::Result<()>;
}
```

We could merge `insert()` and `update()` into a single method but I don't know... I find this a bit cleaner. (not feeling strongly, open to changing my mind though)

Do you perhaps have any other potential breaking changes in mind? Now would be a great chance to think about that :)

> The alternative is to expect sockets and such are already set in non-blocking mode, and bump to 2.0.

👍 A PR would be welcome, thanks!

**Comment by YorickPeterse** [comment]
@stjepang Sorry, what I meant was that the current `Poller::insert` and `Poller::interest` methods are not a problem for me after all. So now API changes for that wowuld be necessary.

**Comment by stjepang** [comment]
But still, if you want to get rid of unnecessary syscalls, isn't it best to pass interest together with EPOLL_CTL_ADD so that we don't *have to* follow up with an EPOLL_CTL_MOD syscall after that?

**Comment by YorickPeterse** [comment]
@stjepang Good point. In that case I agree, fewer system calls would be better. I'll take a look at your proposed setup later tonight and set up a PR.

**Comment by stjepang** [comment]
By the way, I just realized that kqueue allows an important optimization - arbitrary number of `kevent()` syscalls can be squished into one! The kqueue API has a very natural way of batching a bunch of syscalls into one.

**Comment by YorickPeterse** [comment]
@stjepang With this taken care of, I'll set up a draft MR for async-io to make use of these new changes. This will of course fail until a new polling version is released, but it should take some work off your hands :smile_cat:

**Comment by YorickPeterse** [comment]
@stjepang Actually I think there may be an unexpected change: previously calling `Poller.insert` would insert an event using `NOTIFY_KEY'. This constant is not public, so users can't replicate this behaviour using `Event`. async-io in a few places seems to depend on that, at least from my understanding.

What is the best approach here? I think we could add `Event::notify`, but I wonder if that way too implementation details would get exposed. Or perhaps we need to change async-io in ways beyond just find-replacing some method calls.

**Comment by stjepang** [comment]
Don't worry, I got this :)

**Comment by stjepang** [comment]
I published `polling` v2.0.0 and updated `async-io` to use the new version.

**Comment by YorickPeterse** [comment]
@stjepang Thanks!
